### PR TITLE
Migrate from `ethereum-types` to `alloy-primitives`

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

partially #6022 

## Proposed Changes

Migrate LH from `ethereum-types` to `alloy-primitives`

## Additional Info

Some of the dependencies are referencing branches from my repo. I've opened PR's that need to be merged into those repos so we can revert back to referencing the sigp repos.

I've attempted to patch in some dependencies, but quickly faced some cyclic dep issues that I was unable to resolve

### CODE REVIEW

**Comment by michaelsproul** [comment]
I tried investigating the circular dependency issue that crops up when using `[patch]`.

I think it is down to how Cargo handles optional dependencies, and possibly the `ethereum_ssz?/arbitrary` dependency here:

https://github.com/alloy-rs/core/blob/6540523385ea10c9adce31beae7106cfa7da770f/crates/primitives/Cargo.toml#L96-L105

I made a minimal test case here, and it is very broken, even without the `?` syntax:

https://github.com/michaelsproul/circular-test

I think our two options are:

1. Tolerate the existence of _two_ `ethereum_ssz` crates in the build tree until `alloy-primitives` is updated to not depend on `ethereum_ssz` _at all_. If we bump the `ethereum_ssz` version to `0.6` on the alloy branch, and depend on that version in Lighthouse, then the `ethereum_ssz` dep of `alloy-primitives` will not get patched.
2. Raise a PR to `alloy-primitives` to remove the `ethereum_ssz` dep, and make this part of the `patch` section. This probably needs to be the long-term solution, so maybe we should start working on this even if we take option (1) for now.

**Comment by michaelsproul** [comment]
Related Cargo issue:

- https://github.com/rust-lang/cargo/issues/6743

**Comment by eserilev** [comment]
Thanks for taking a look at this @michaelsproul 

For option (2), removing the `ethereum_ssz` dep, I think this means we'll need to impl ssz Encode & Decode directly in the `alloy_primitives` crate. Does that sound correct?

**Comment by michaelsproul** [comment]
I think we should move the SSZ impls that are currently in alloy-primitives into ethereum_ssz itself.

E.g. implementing Decode/Encode for H256 should be in ethereum_ssz

**Comment by eserilev** [comment]
I went ahead and made some progress on option (2)

- PR to `alloy` https://github.com/alloy-rs/core/pull/701
- PR to `ethereum_ssz` https://github.com/sigp/ethereum_ssz/pull/26

I've also patched in the relevant deps into Lighthouse. No more circular dependency issues!

**Comment by michaelsproul** [comment]
I've released preliminary versions of all our crates which we can use in this PR. I think we could merge it to `unstable` and then come back for the coordinated release to remove `ethereum_ssz 0.5.4` later.

**Comment by eserilev** [comment]
sounds good, will resolve the conflicts and get those dependencies pointed to the right versions

**Comment by eserilev** [comment]
Updated to point to the correct dependencies and resolved all conflicts. Since we're not patching any deps in anymore, we're actually ok to use the existing alloy-primitives crate.

File: beacon_node/network/src/service.rs
Comment (eserilev) [inline beacon_node/network/src/service.rs]: this can panic if the value is larger than what can fit in u256. that should never happen here, but flagging it just in case. theres a few other places in this PR where we've done things similarly.

**Comment by michaelsproul** [comment]
I've just released new versions of all the packages depending on `alloy-primitives 0.8`, so we should be able to cut out the duplicate version of `ethereum_ssz` now!

Can you bump every dep by +0.1.0?

**Comment by eserilev** [comment]
we cant bump alloy-primitives to 0.8 until this gets merged: https://github.com/alloy-rs/alloy/pull/1167

`alloy-consensus` still depends on `alloy-primitives` 0.7.7 and theres some diff between `alloy_primitives::TxKind` on 0.7 vs 0.8 that causes issues

maybe theres a way to work around this issue? not 100% sure. We can either wait for the new `alloy` release or figure out a workaround

EDIT: in 
11d8a26 I've commented out the problematic code and added a TODO to reintroduce it once `alloy` cuts a new release

File: consensus/types/src/attestation.rs
Comment (eserilev) [inline consensus/types/src/attestation.rs:415]: Snuck in this small change since the new `ssz_types` release has this function available

**Comment by mattsse** [comment]
fyi new alloy-rs/alloy crate release coming later today, tmrw at the latest

**Comment by eserilev** [comment]
alloy cut a release so were no longer blocked by that

File: beacon_node/beacon_chain/src/metrics.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/metrics.rs]: I'm a little bit hesitant about `get_i64` and similar functions because they can panic.

For use cases like metrics I would prefer something that might truncate or return junk if we make a mistake.

Maybe we could adds something like `to_low_u64_le()` to the mixin trait?

File: beacon_node/execution_layer/src/lib.rs
Comment (michaelsproul) [inline beacon_node/execution_layer/src/lib.rs:2201]: I think we probably want to keep this condition, or incorporate something similar into the safe mixin method for converting H256/U256 to i64

File: beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs
Comment (michaelsproul) [inline beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs]: We might want big-endian here, yeah?

File: beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs
Comment (michaelsproul) [inline beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs]: Big-endian again?

I'm actually wondering if a better API would be to pass `[u8; 32]` to these compute functions, and do the endianness conversion to `U256` inside.

File: beacon_node/network/src/service.rs
Comment (michaelsproul) [reply beacon_node/network/src/service.rs]: this is another LE

File: beacon_node/network/src/service/tests.rs
Comment (michaelsproul) [inline beacon_node/network/src/service/tests.rs]: another LE (I think lets do the `[u8; 32]` API and that will force a refactor of all of these)

File: common/eth2/src/types.rs
Comment (michaelsproul) [inline common/eth2/src/types.rs]: To maintain the same behaviour I think we should use `from_str_radix(s, 10)` here. The `from_str` function accepts hex, octal and binary inputs which we should probably reject:

https://docs.rs/ruint/1.12.3/src/ruint/string.rs.html#108-121

File: common/eth2/src/types.rs
Comment (michaelsproul) [inline common/eth2/src/types.rs]: As above.

File: consensus/types/src/execution_transaction.rs
Comment (michaelsproul) [inline consensus/types/src/execution_transaction.rs:1]: This file can be deleted

File: crypto/bls/src/lib.rs
Comment (michaelsproul) [inline crypto/bls/src/lib.rs]: Rather than creating 3 copies of this trait, maybe we should pull it out into its own crate in `consensus` which is pulled in by `types`, `merkle_proof` and `bls`?

We could call the crate `fixed_bytes`

File: testing/ef_tests/src/cases/epoch_processing.rs
Comment (michaelsproul) [inline testing/ef_tests/src/cases/epoch_processing.rs]: This comment can be removed as the tests pass now

File: testing/ef_tests/src/cases/fork_choice.rs
Comment (michaelsproul) [inline testing/ef_tests/src/cases/fork_choice.rs]: May as well avoid this newline

```suggestion
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looks great on the whole!

I just have a couple of suggestions and then we can merge!

File: beacon_node/execution_layer/src/lib.rs
Comment (eserilev) [reply beacon_node/execution_layer/src/lib.rs:2201]: Introduced a safe mixin here: [e555728](https://github.com/sigp/lighthouse/pull/6078/commits/e555728d25bb960595f4eef6f64ea074317921ad)

File: beacon_node/beacon_chain/src/metrics.rs
Comment (eserilev) [reply beacon_node/beacon_chain/src/metrics.rs]: Introduced `to_low_u64_le()` here: [4cf9de8](https://github.com/sigp/lighthouse/pull/6078/commits/4cf9de815d4d24e802d07677318c388f72adc0a2)

File: common/eth2/src/types.rs
Comment (eserilev) [reply common/eth2/src/types.rs]: a7876e7

File: common/eth2/src/types.rs
Comment (eserilev) [reply common/eth2/src/types.rs]: a7876e7

File: crypto/bls/src/lib.rs
Comment (eserilev) [reply crypto/bls/src/lib.rs]: Implemented in [5162316](https://github.com/sigp/lighthouse/pull/6078/commits/5162316d8927b69451837468c211c69576a1fbf3)

File: testing/ef_tests/src/cases/epoch_processing.rs
Comment (eserilev) [reply testing/ef_tests/src/cases/epoch_processing.rs]: deleted TODO

File: testing/ef_tests/src/cases/fork_choice.rs
Comment (eserilev) [reply testing/ef_tests/src/cases/fork_choice.rs]: deleted newline

File: consensus/types/src/execution_transaction.rs
Comment (eserilev) [reply consensus/types/src/execution_transaction.rs:1]: deleted

File: beacon_node/network/src/service/tests.rs
Comment (eserilev) [reply beacon_node/network/src/service/tests.rs]: updated: 
bbeb7de

File: beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs
Comment (eserilev) [reply beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs]: updated api here:  bbeb7de

File: beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs
Comment (eserilev) [reply beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs]: bbeb7de

**Comment by eserilev** [comment]
this should be ready for another review!

File: Cargo.toml
Comment (michaelsproul) [inline Cargo.toml:208]: I think this also needs to be added to `[workspace.members]` above, but I'm not 100% on that

**Review by michaelsproul — APPROVED** [review]

**Comment by michaelsproul** [comment]
@mergify queue

**Pull request merged**
